### PR TITLE
REF: Update PtychoNN integration to 0.3.x API

### PR DESCRIPTION
**Pull request by carterbox**
Migrates the PtychoNN integration to the 0.3.x API. This new PtychoNN API is functional instead of object oriented and uses Pytorch Lighting under the hood.

## TODOS
- [x] Reduce precision of "Validation Set Fraction Size" slider to `0.1`. We don't need to be able to choose precision to 1/1000th of a dataset
- [x] Remove "Optimization Epoch Per Half Cycle" field. It's not longer user settable.
- [x] Redirect "Status Interval" field to Trainer(log_every_n_steps=50). May need to adjust PtychoNN API.
- [x] Change UI section under "Save Training Artifacts". Path is a folder, and Suffix is the name of the model checkpoint/parameters inside that folder. It's unclear that Suffix must end in `.pth` in order for the model state to be loaded again by Ptychodus
- [x] Expose method for changing data parallelism model to single process
  - https://pytorch-lightning.readthedocs.io/en/0.9.0/multi_gpu.html#distributed-modes 
  - https://github.com/mcherukara/PtychoNN/pull/29

### CODE REVIEW

File: ptychodus/model/ptychonn/reconstructor.py [hunk 1]
```diff
@@ -105,12 +104,24 @@ def __init__(self, modelSettings: PtychoNNModelSettings,
...
+            parameters = None
+        else:
+            path = None
+            parameters = dict(
+                nconv=self._modelSettings.numberOfConvolutionKernels.value,
+                use_batch_norm=self._modelSettings.useBatchNormalization.value,
+                enable_amplitude=self._enableAmplitude,
+                max_lr=float(self._trainingSettings.maximumLearningRate.value),
+                min_lr=float(self._trainingSettings.minimumLearningRate.value),
+            )
```
Comment (carterbox) [inline ptychodus/model/ptychonn/reconstructor.py:120]: The behavior of createModel is now that a model is created if the stateFilePath does not exist, otherwise it tries to load the model from the existing file. Not sure if that was the previous behavior. This means that a model is always returned, but if the user provided an incorrect model path, they will get a randomly initialized model without any error.

**Comment by carterbox** [comment]
@stevehenke, this PR needs some help with the TODO list. I believe that the training works, but there are some GUI elements that are now awkward/obsolete, and I don't know where the implementation for those UI elements are located.

**Comment by carterbox** [comment]
It looks like `dp` is no longer a distributed training option for pytorch lighting. The options are listed [here](https://lightning.ai/docs/pytorch/stable/extensions/strategy.html#selecting-a-built-in-strategy).

Perhaps `ddp_spawn` might be the closest to what you are looking for?

**Review by stevehenke — APPROVED** [review]

**Comment by carterbox** [comment]
Importing this this version fails if PtychoNN is not installed because you have imported pytorch lightning directly in order to use type hints. I believe this is called an API leak? Ptychodus is supposed to be agnostic to the PtychoNN backend, but since PtychoNN references lighting classes in its public API, Ptychodus needs to know about these classes. Probably I should reexport lightning.Trainer as ptychonn.Trainer such that a direct import of lighting is not needed by Ptychodus.

File: ptychodus/model/ptychonn/model.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+from pathlib import Path
+import logging
+
+import lightning
+import ptychonn
```
Comment (carterbox) [inline ptychodus/model/ptychonn/model.py:5]: Both of these imports cause ptychonn to become a dependency for the base ptychodus package. I'm not sure how to avoid these imports, but ptychonn should not be a requirement to run ptychodus.
